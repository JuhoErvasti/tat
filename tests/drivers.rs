use tat::{fixtures::{datasets::*, init_table, TatTestStructure, TatTestUtils}, table::TatTable};
use rstest::rstest;
use insta::assert_snapshot;

fn test_single_layer_dataset(driver_name: &str, structs: (TatTestStructure, TatTable)) {
    let (test, mut t) = structs;
    TatTestUtils::set_layer_index_and_update(0, &mut t, &test.tatevent_rx);
    TatTestUtils::prepare_table_attribute_view_all_features(&t, &test.tatevent_rx, test.ds_request_tx.clone());

    let snapshot_name = format!("{}_{}", driver_name, t.layer_schema().unwrap().name());
    assert_snapshot!(snapshot_name, t);
}

fn test_all_layers(driver_name: &str, structs: (TatTestStructure, TatTable)) {
    let (test, mut t) = structs;
    for index in 0..t.layer_schemas().len() as usize {
        TatTestUtils::set_layer_index_and_update(index, &mut t, &test.tatevent_rx);
        TatTestUtils::prepare_table_attribute_view_all_features(&t, &test.tatevent_rx, test.ds_request_tx.clone());

        assert_eq!(t.layer_schema().unwrap().index(), index);
        let snapshot_name = format!("{}_{}", driver_name, t.layer_schema().unwrap().name());
        assert_snapshot!(snapshot_name, t);
    }

    test.terminate();
}

#[rstest]
fn test_basic_gdb(basic_gdb: TatTestStructure) {
    test_all_layers("openfilegdb", init_table(basic_gdb));
}

#[rstest]
fn test_basic_gml(basic_gml: TatTestStructure) {
    test_all_layers("gml", init_table(basic_gml));
}

#[rstest]
fn test_basic_jsonfg(basic_jsonfg: TatTestStructure) {
    test_all_layers("jsonfg", init_table(basic_jsonfg));
}

#[rstest]
fn test_basic_mapml(basic_mapml: TatTestStructure) {
    test_all_layers("mapml", init_table(basic_mapml));
}

#[rstest]
fn test_basic_ods(basic_ods: TatTestStructure) {
    test_all_layers("ods", init_table(basic_ods));
}

// the reason these are individual functions like this is because it works
// better with the insta snapshots
#[rstest]
fn test_basic_shp_point(basic_shp_point: TatTestStructure) {
    test_single_layer_dataset("esri_shapefile", init_table(basic_shp_point));
}

#[rstest]
fn test_basic_shp_line(basic_shp_line: TatTestStructure) {
    test_single_layer_dataset("esri_shapefile", init_table(basic_shp_line));
}

#[rstest]
fn test_basic_shp_polygon(basic_shp_polygon: TatTestStructure) {
    test_single_layer_dataset("esri_shapefile", init_table(basic_shp_polygon));
}

#[rstest]
fn test_basic_shp_multipolygon(basic_shp_multipolygon: TatTestStructure) {
    test_single_layer_dataset("esri_shapefile", init_table(basic_shp_multipolygon));
}

#[rstest]
fn test_basic_shp_nogeom(basic_shp_nogeom: TatTestStructure) {
    test_single_layer_dataset("esri_shapefile", init_table(basic_shp_nogeom));
}

#[rstest]
fn test_basic_csv_point(basic_csv_point: TatTestStructure) {
    test_single_layer_dataset("csv", init_table(basic_csv_point));
}

#[rstest]
fn test_basic_csv_line(basic_csv_line: TatTestStructure) {
    test_single_layer_dataset("csv", init_table(basic_csv_line));
}

#[rstest]
fn test_basic_csv_polygon(basic_csv_polygon: TatTestStructure) {
    test_single_layer_dataset("csv", init_table(basic_csv_polygon));
}

#[rstest]
fn test_basic_csv_multipolygon(basic_csv_multipolygon: TatTestStructure) {
    test_single_layer_dataset("csv", init_table(basic_csv_multipolygon));
}

#[rstest]
fn test_basic_csv_nogeom(basic_csv_nogeom: TatTestStructure) {
    test_single_layer_dataset("csv", init_table(basic_csv_nogeom));
}

#[rstest]
fn test_basic_geojson_point(basic_geojson_point: TatTestStructure) {
    test_single_layer_dataset("geojson", init_table(basic_geojson_point));
}

#[rstest]
fn test_basic_geojson_line(basic_geojson_line: TatTestStructure) {
    test_single_layer_dataset("geojson", init_table(basic_geojson_line));
}

#[rstest]
fn test_basic_geojson_polygon(basic_geojson_polygon: TatTestStructure) {
    test_single_layer_dataset("geojson", init_table(basic_geojson_polygon));
}

#[rstest]
fn test_basic_geojson_multipolygon(basic_geojson_multipolygon: TatTestStructure) {
    test_single_layer_dataset("geojson", init_table(basic_geojson_multipolygon));
}

#[rstest]
fn test_basic_geojson_nogeom(basic_geojson_nogeom: TatTestStructure) {
    test_single_layer_dataset("geojson", init_table(basic_geojson_nogeom));
}

#[rstest]
fn test_basic_geojsonseq_point(basic_geojsonseq_point: TatTestStructure) {
    test_single_layer_dataset("geojsonseq", init_table(basic_geojsonseq_point));
}

#[rstest]
fn test_basic_geojsonseq_line(basic_geojsonseq_line: TatTestStructure) {
    test_single_layer_dataset("geojsonseq", init_table(basic_geojsonseq_line));
}

#[rstest]
fn test_basic_geojsonseq_polygon(basic_geojsonseq_polygon: TatTestStructure) {
    test_single_layer_dataset("geojsonseq", init_table(basic_geojsonseq_polygon));
}

#[rstest]
fn test_basic_geojsonseq_multipolygon(basic_geojsonseq_multipolygon: TatTestStructure) {
    test_single_layer_dataset("geojsonseq", init_table(basic_geojsonseq_multipolygon));
}

#[rstest]
fn test_basic_geojsonseq_nogeom(basic_geojsonseq_nogeom: TatTestStructure) {
    test_single_layer_dataset("geojsonseq", init_table(basic_geojsonseq_nogeom));
}

#[rstest]
fn test_basic_jml_point(basic_jml_point: TatTestStructure) {
    test_single_layer_dataset("jml", init_table(basic_jml_point));
}

#[rstest]
fn test_basic_jml_line(basic_jml_line: TatTestStructure) {
    test_single_layer_dataset("jml", init_table(basic_jml_line));
}

#[rstest]
fn test_basic_jml_polygon(basic_jml_polygon: TatTestStructure) {
    test_single_layer_dataset("jml", init_table(basic_jml_polygon));
}

#[rstest]
fn test_basic_jml_multipolygon(basic_jml_multipolygon: TatTestStructure) {
    test_single_layer_dataset("jml", init_table(basic_jml_multipolygon));
}

#[rstest]
fn test_basic_jml_nogeom(basic_jml_nogeom: TatTestStructure) {
    test_single_layer_dataset("jml", init_table(basic_jml_nogeom));
}

#[rstest]
fn test_basic_mapinfofile_point(basic_mapinfofile_point: TatTestStructure) {
    test_single_layer_dataset("mapinfo_file", init_table(basic_mapinfofile_point));
}

#[rstest]
fn test_basic_mapinfofile_line(basic_mapinfofile_line: TatTestStructure) {
    test_single_layer_dataset("mapinfo_file", init_table(basic_mapinfofile_line));
}

#[rstest]
fn test_basic_mapinfofile_polygon(basic_mapinfofile_polygon: TatTestStructure) {
    test_single_layer_dataset("mapinfo_file", init_table(basic_mapinfofile_polygon));
}

#[rstest]
fn test_basic_mapinfofile_multipolygon(basic_mapinfofile_multipolygon: TatTestStructure) {
    test_single_layer_dataset("mapinfo_file", init_table(basic_mapinfofile_multipolygon));
}

#[rstest]
fn test_basic_mapinfofile_nogeom(basic_mapinfofile_nogeom: TatTestStructure) {
    test_single_layer_dataset("mapinfo_file", init_table(basic_mapinfofile_nogeom));
}

#[rstest]
fn test_basic_xlsx_point(basic_xlsx_point: TatTestStructure) {
    test_single_layer_dataset("xlsx", init_table(basic_xlsx_point));
}

#[rstest]
fn test_basic_xlsx_line(basic_xlsx_line: TatTestStructure) {
    test_single_layer_dataset("xlsx", init_table(basic_xlsx_line));
}

#[rstest]
fn test_basic_xlsx_polygon(basic_xlsx_polygon: TatTestStructure) {
    test_single_layer_dataset("xlsx", init_table(basic_xlsx_polygon));
}

#[rstest]
fn test_basic_xlsx_multipolygon(basic_xlsx_multipolygon: TatTestStructure) {
    test_single_layer_dataset("xlsx", init_table(basic_xlsx_multipolygon));
}

#[rstest]
fn test_basic_xlsx_nogeom(basic_xlsx_nogeom: TatTestStructure) {
    test_single_layer_dataset("xlsx", init_table(basic_xlsx_nogeom));
}
