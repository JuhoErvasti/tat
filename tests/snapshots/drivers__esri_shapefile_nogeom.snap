---
source: tests/drivers.rs
expression: result
---
text_field,i32_field,i64_field,decimal_fi,date_field,datetime_f,bool_field,blob_field,json_field
text,10,100,1.541000000000000,1970/07/10,2025-07-19T20:45:45Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
buzz,374,-1384,-104.092138000000006,2010/05/25,2016-05-03T01:40:15.048Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
check,173,2474,-170.296267000000000,2009/02/22,2015-12-13T04:04:17.556Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
die,681,-404,7.189099000000000,2028/04/06,2023-12-28T17:57:16.036Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
participate,967,-3170,72.283052999999995,2022/02/13,2025-12-12T02:11:45.662Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
huge,980,-403,90.188894000000005,2013/06/29,2018-01-30T07:39:08.074Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
read,295,-1531,-84.884591999999998,2023/08/27,2022-01-13T19:28:32.476Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
check,894,301,165.484084999999993,2024/06/07,2024-03-09T14:32:41.238Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
strive,841,-1536,43.659255000000002,2020/12/29,2029-01-26T07:53:25.466Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
strive,729,4101,-192.625146000000001,2018/12/22,2019-12-23T21:51:52.335Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
fall,456,3834,-152.688532000000009,2029/04/17,2016-07-04T06:13:35.176Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
participate,595,-219,56.665858000000000,2016/06/23,2009-09-10T01:26:32.349Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
study,518,4042,14.859641999999999,2028/12/22,2015-03-22T08:33:33.770Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
respect,577,3366,5.526759000000000,2012/11/27,2019-01-02T10:08:57.047Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
die,434,2946,164.085164999999989,2025/06/11,2026-06-17T07:23:28.466Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
verify,-28,1443,45.941403000000001,2027/03/03,2008-06-10T19:20:09.165Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
study,206,2529,-129.470352999999989,2009/06/05,2009-12-15T09:45:58.501Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
verify,540,-3621,70.006603999999996,2026/02/26,2022-01-21T08:04:41.328Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
verify,-92,2104,85.831181000000001,2021/03/05,2022-04-05T02:29:18.756Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
murky,163,3234,59.299899000000003,2016/11/23,2012-07-28T01:56:33.320Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
arrest,670,2865,-165.115901000000008,2019/04/19,2026-11-27T03:46:43.114Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
participate,261,204,-90.869298999999998,2013/07/16,2019-11-13T17:56:37.309Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
loving,615,4905,-135.931366999999995,2024/06/12,2015-01-11T10:11:40.443Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
breath,873,-3744,23.026824999999999,2009/06/17,2021-06-18T19:26:26.856Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
breath,507,-1789,181.181896999999992,2026/02/08,2011-04-23T12:40:14.519Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
read,134,307,-115.438984000000005,2008/02/02,2029-01-05T19:17:11.004Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
breath,177,-4535,-20.305130999999999,2025/06/03,2017-05-24T17:41:15.977Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
participate,400,2477,57.760561000000003,2023/09/04,2009-11-14T14:39:00.661Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
tremble,-7,-4141,163.773981999999990,2013/02/20,2023-05-15T05:32:31.008Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
study,873,-3609,45.920408999999999,2011/07/05,2027-03-31T08:35:38.187Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
raspy,842,4792,95.786027000000004,2015/06/20,2021-05-14T10:08:54.558Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
current,323,-1861,-82.259782000000001,2019/09/02,2020-01-20T22:50:05.982Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
bloody,418,4161,125.614895000000004,2014/06/16,2019-01-07T06:42:48.012Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
murky,918,-4092,196.367059000000012,2021/06/24,2025-01-11T09:39:58.739Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
current,212,4349,106.562074999999993,2026/12/26,2011-06-16T13:44:14.665Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
read,-51,-425,154.331480999999997,2026/06/09,2013-10-12T12:47:54.145Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
verify,870,-3350,73.297849999999997,2013/01/29,2011-04-02T17:39:28.947Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
loving,13,2095,98.944012999999998,2017/09/20,2028-12-11T02:01:38.787Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
verify,356,-4767,-103.234727000000007,2008/08/18,2016-10-09T16:58:43.109Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
die,3,-3577,-61.458798999999999,2021/08/13,2023-01-07T15:59:55.057Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
raspy,634,-113,-13.152386999999999,2019/01/16,2027-04-28T21:35:52.459Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
murky,306,3620,-1.471470000000000,2021/11/25,2016-03-15T04:34:20.185Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
bad,987,2034,15.920553999999999,2009/09/08,2022-08-28T19:16:38.777Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
courageous,584,-90,163.497059000000007,2023/03/24,2010-05-29T07:45:13.766Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
courageous,572,536,-0.134892000000000,2025/07/27,2007-12-28T04:04:33.089Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
murky,321,3836,-29.019805999999999,2013/09/14,2020-05-20T20:01:18.398Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
current,946,1786,-75.821284000000006,2024/01/23,2015-06-29T17:44:11.773Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
bad,906,-4961,58.555764000000003,2021/09/28,2017-05-19T03:07:07.501Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
fall,655,4850,62.768847000000001,2022/01/25,2025-09-28T19:17:40.143Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
fall,847,-4739,-189.710329999999999,2013/01/20,2028-10-30T08:16:35.507Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
scrawny,227,4535,8.927695000000000,2019/06/29,2025-09-12T14:08:51.147Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
huge,-12,-4563,-160.145694999999989,2007/11/27,2027-11-03T05:32:52.387Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
courageous,731,-2645,-33.973891000000002,2009/04/14,2013-02-14T11:24:52.293Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
read,100,-1557,-28.323145000000000,2011/03/13,2013-07-16T09:32:22.022Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
study,56,-674,-162.842936000000009,2021/02/21,2010-06-28T16:41:23.595Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
bad,702,-198,140.328748999999988,2022/03/20,2026-04-14T08:38:41.864Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
current,430,-3908,-157.329372000000006,2013/09/27,2024-12-29T02:45:42.003Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
huge,895,1686,53.586337999999998,2025/10/17,2018-10-12T13:54:14.202Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
read,-18,736,41.903661000000000,2025/05/08,2014-02-23T16:35:14.647Z,T,626C6F620A,{"another_key":"another_value","key":"value"}
legal,397,-684,166.452107000000012,2012/08/19,2020-04-12T16:57:19.989Z,F,626C6F620A,{"another_key":"another_value","key":"value"}
